src/Shared/Model/ArtifactUri.cs

```diff
@@ -0,0 +1,85 @@
+// Copyright (c) Microsoft Corporation. Licensed under the MIT License.
+
+namespace Microsoft.CST.OpenSource.Model;
+
+using PackageManagers;
+using Polly;
+using Polly.Contrib.WaitAndRetry;
+using Polly.Retry;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Net;
+using System.Net.Http;
+using System.Threading.Tasks;
+
+/// <summary>
+/// A record to represent the type, uri, and extension for an artifact associated with a package.
+/// </summary>
+/// <typeparam name="T">The enum to represent the artifact type.</typeparam>
+public record ArtifactUri<T> where T : Enum
+{
+    /// <summary>
+    /// Initializes a new instance of <see cref="ArtifactUri{T}"/>.
+    /// </summary>
+    /// <param name="type">The type of artifact for this <see cref="ArtifactUri{T}"/>.</param>
+    /// <param name="uri">The <see cref="Uri"/> this artifact can be found at.</param>
+    /// <param name="extension">The file extension for the file found at the <paramref name="uri"/>.</param>
+    public ArtifactUri(T type, Uri uri, string? extension = null)
+    {
+        Type = type;
+        Uri = uri;
+        Extension = extension ?? Path.GetExtension(uri.AbsolutePath);
+    }
+
+    /// <summary>
+    /// Initializes a new instance of <see cref="ArtifactUri{T}"/>.
+    /// </summary>
+    /// <param name="type">The type of artifact for this <see cref="ArtifactUri{T}"/>.</param>
+    /// <param name="uri">The string of the uri this artifact can be found at.</param>
+    /// <param name="extension">The file extension for the file found at the <paramref name="uri"/>.</param>
+    public ArtifactUri(T type, string uri, string? extension = null) : this(type, new Uri(uri), extension) { }
+
+    /// <summary>
+    /// The enum representing the artifact type for the project manager associated with this artifact.
+    /// </summary>
+    public T Type { get; }
+
+    /// <summary>
+    /// The <see cref="Uri"/> for where this artifact can be found online.
+    /// </summary>
+    public Uri Uri { get; }
+
+    /// <summary>
+    /// The file extension for this artifact file.
+    /// </summary>
+    public string Extension { get; }
```

Comment:
```suggestion
    public string Extension => Path.GetExtension(uri.AbsolutePath);
```
I think its better if these cannot be out of sync